###  Загрузить библиотеки

In [4]:
import numpy as np  #№
import pandas as pd  ##
from sklearn.neighbors import NearestNeighbors
import seaborn as sns
from operator import itemgetter   ##
from sklearn.metrics.pairwise import cosine_similarity
#from surprise import Dataset, Reader, KNNBasic
#from surprise.model_selection import train_test_split

## Загрузить датасет с играми

In [28]:
steam_games = pd.read_json('steam_games.json')

In [29]:
steam_games

10       \
appid                                                             10   
name                                                  Counter-Strike   
short_description  Play the world's number 1 online action game. ...   
developer                                                      Valve   
publisher                                                      Valve   
genre                                                         Action   
tags               {'Action': 5426, 'FPS': 4831, 'Multiplayer': 3...   
type                                                            game   
categories         [Multi-player, Valve Anti-Cheat enabled, Onlin...   
owners                                      10,000,000 .. 20,000,000   
positive                                                      201215   
negative                                                        5199   
price                                                            999   
initialprice                                                     999   
discount                                                           0   
ccu                                                            13990   
languages          English, French, German, Italian, Spanish - Sp...   
platforms              {'windows': True, 'mac': True, 'linux': True}   
release_date                                               2000/11/1   
required_age                                                       0   
website                                                                
header_image       https://cdn.akamai.steamstatic.com/steam/apps/...   

                                                             1000000  \
appid                                                        1000000   
name                                                       ASCENXION   
short_description  ASCENXION is a 2D shoot 'em up game where you ...   
developer                                     IndigoBlue Game Studio   
publisher                                   PsychoFlux Entertainment   
genre                                       Action, Adventure, Indie   
tags               {'Shoot 'Em Up': 186, 'Metroidvania': 181, 'Bu...   
type                                                            game   
categories         [Single-player, Partial Controller Support, St...   
owners                                                   0 .. 20,000   
positive                                                          27   
negative                                                           5   
price                                                            999   
initialprice                                                     999   
discount                                                           0   
ccu                                                                0   
languages                        English, Korean, Simplified Chinese   
platforms            {'windows': True, 'mac': False, 'linux': False}   
release_date                                              2021/05/14   
required_age                                                       0   
website                                                                
header_image       https://cdn.akamai.steamstatic.com/steam/apps/...   

                                                             1000010  \
appid                                                        1000010   
name                                                     Crown Trick   
short_description  Enter a labyrinth that moves as you move, wher...   
developer                                               NEXT Studios   
publisher                                       Team17, NEXT Studios   
genre                                Adventure, Indie, RPG, Strategy   
tags               {'Rogue-like': 268, 'Turn-Based Combat': 254, ...   
type                                                            game   
categories         [Single-player, Partial Controller Support, St...   
owners                     

In [30]:
steam_games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22 entries, appid to header_image
Columns: 55691 entries, 10 to 999990
dtypes: object(55691)
memory usage: 9.3+ MB


In [31]:
steam_games = pd.DataFrame(steam_games.transpose())
steam_games

appid                            name  \
10            10                  Counter-Strike   
1000000  1000000                       ASCENXION   
1000010  1000010                     Crown Trick   
1000030  1000030     Cook, Serve, Delicious! 3?!   
1000040  1000040                            细胞战争   
...          ...                             ...   
999880    999880     NASA's Exoplanet Excursions   
999890    999890                          Bruken   
999900    999900               Studio by RADiCAL   
999930    999930  Fantasy Sino-Japanese War 幻想甲午   
999990    999990                   Bouncing Hero   

                                         short_description  \
10       Play the world's number 1 online action game. ...   
1000000  ASCENXION is a 2D shoot 'em up game where you ...   
1000010  Enter a labyrinth that moves as you move, wher...   
1000030  Cook, serve and manage your food truck as you ...   
1000040  这是一款打击感十足的细胞主题游戏！操作简单但活下去却不简单，“你”作为侵入人体的细菌病毒，通...   
...                                                    ...   
999880   Take VR journeys through the amazing TRAPPIST-...   
999890   Explore a large spherical world, enjoying all ...   
999900   RADiCAL Studio is a real-time, unlimited motio...   
999930   You unfortunately passed through the Sino-Japa...   
999990   Bouncing Hero is an extreme auto-jump platform...   

                                              developer  \
10                                                Valve   
1000000                          IndigoBlue Game Studio   
1000010                                    NEXT Studios   
1000030                             Vertigo Gaming Inc.   
1000040                                   DoubleC Games   
...                                                 ...   
999880   Caltech/IPAC, NASA's Jet Propulsion Laboratory   
999890                                   FlairBot Games   
999900                                          RADiCAL   
999930                                           张八万工作室   
999990                                   ZPink, zniq.co   

                                              publisher  \
10                                                Valve   
1000000                        PsychoFlux Entertainment   
1000010                            Team17, NEXT Studios   
1000030                             Vertigo Gaming Inc.   
1000040                                   DoubleC Games   
...                                                 ...   
999880   Caltech/IPAC, NASA's Jet Propulsion Laboratory   
999890                                   FlairBot Games   
999900                           RADiCAL Solutions, LLC   
999930                                           张八万工作室   
999990                         PsychoFlux Entertainment   

                                       genre  \
10                                    Action   
1000000             Action, Adventure, Indie   
1000010      Adventure, Indie, RPG, Strategy   
1000030  Action, Indie, Simulation, Strategy   
1000040    Action, Casual, Indie, Simulation   
...                                      ...   
999880                             Education   
999890              Adventure, Casual, Indie   
999900                  Animation & Modeling   
999930    Indie, RPG, Strategy, Early Access   
999990                 Action, Casual, Indie   

                                                      tags  type  \
10       {'Action': 5426, 'FPS': 4831, 'Multiplayer': 3...  game   
1000000  {'Shoot 'Em Up': 186, 'Metroidvania': 181, 'Bu...  game   
1000010  {'Rogue-like': 268, 'Turn-Based Combat': 254, ...  game   
1000030  {'Typing': 221, 'Management': 213, 'Casual': 2...  game   
1000040  {'Action': 22, 'Casual': 22, 'Indie': 21, 'Sim...  game   
...                                                    ...   ...   
999880              {'Education': 22, 'VR': 6, 'Space': 5}  game   
999890        {'Casual': 21, 'Indie': 21, 'Adventure': 20}  game   
999900              {'Animation & 

## Убрать игры с низким количеством пользователей

In [32]:
steam_games[steam_games.columns[9]].unique() # '0 .. 20,000','20,000 .. 50,000','50,000 .. 100,000'

array(['10,000,000 .. 20,000,000', '0 .. 20,000', '200,000 .. 500,000',
       '100,000 .. 200,000', '20,000 .. 50,000', '50,000 .. 100,000',
       '500,000 .. 1,000,000', '2,000,000 .. 5,000,000',
       '5,000,000 .. 10,000,000', '1,000,000 .. 2,000,000',
       '20,000,000 .. 50,000,000', '50,000,000 .. 100,000,000',
       '200,000,000 .. 500,000,000'], dtype=object)

In [33]:
values_to_remove = ['0 .. 20,000','20,000 .. 50,000','50,000 .. 100,000']
#df = df[~df['col2'].isin(values_to_remove)]
steam_games = steam_games[~steam_games[steam_games.columns[9]].isin(values_to_remove)]
steam_games

appid                         name  \
10            10               Counter-Strike   
1000010  1000010                  Crown Trick   
1000030  1000030  Cook, Serve, Delicious! 3?!   
1000080  1000080                      Zengeon   
1000360  1000360                Hellish Quart   
...          ...                          ...   
99900      99900               Spiral Knights   
99910      99910               Puzzle Pirates   
999220    999220             Amnesia: Rebirth   
999730    999730         Secret Neighbor Beta   
999860    999860               Enemy On Board   

                                         short_description  \
10       Play the world's number 1 online action game. ...   
1000010  Enter a labyrinth that moves as you move, wher...   
1000030  Cook, serve and manage your food truck as you ...   
1000080  Zengeon is an anime infused Action RPG and Rog...   
1000360  A physics-based, realistic, 3D sword dueling g...   
...                                                    ...   
99900    Join the ranks of the Spiral Knights. Stranded...   
99910    Every activity in Puzzle Pirates is a uniquely...   
999220   A first-person horror adventure. Uncover your ...   
999730   Beta Runs Aug 2-19! A group of kids tries to s...   
999860   Online (6v2) social deduction game that will r...   

                         developer             publisher  \
10                           Valve                 Valve   
1000010               NEXT Studios  Team17, NEXT Studios   
1000030        Vertigo Gaming Inc.   Vertigo Gaming Inc.   
1000080         IndieLeague Studio              2P Games   
1000360                     Kubold                Kubold   
...                            ...                   ...   
99900                  Grey Havens           Grey Havens   
99910                  Grey Havens           Grey Havens   
999220            Frictional Games      Frictional Games   
999730   Hologryph, Dynamic Pixels             tinyBuild   
999860              Windwalk Games        Windwalk Games   

                                                     genre  \
10                                                  Action   
1000010                    Adventure, Indie, RPG, Strategy   
1000030                Action, Indie, Simulation, Strategy   
1000080                      Action, Adventure, Indie, RPG   
1000360                               Action, Early Access   
...                                                    ...   
99900    Action, Adventure, Casual, Free to Play, Indie...   
99910    Adventure, Casual, Free to Play, Massively Mul...   
999220                                    Adventure, Indie   
999730                            Action, Adventure, Indie   
999860   Action, Casual, Free to Play, Indie, Early Access   

                                                      tags  type  \
10       {'Action': 5426, 'FPS': 4831, 'Multiplayer': 3...  game   
1000010  {'Rogue-like': 268, 'Turn-Based Combat': 254, ...  game   
1000030  {'Typing': 221, 'Management': 213, 'Casual': 2...  game   
1000080  {'Action': 138, 'Indie': 128, 'RPG': 121, 'Adv...  game   
1000360  {'Early Access': 222, '3D Fighter': 201, 'Swor...  game   
...                                                    ...   ...   
99900    {'Free to Play': 1167, 'Massively Multiplayer'...  game   
99910    {'Free to Play': 117, 'Massively Multiplayer':...  game   
999220   {'Horror': 241, 'Survival Horror': 197, 'Psych...  game   
999730   {'Multiplayer': 57, 'Horror': 54, 'Free to Pla...  game   
999860   {'Free to Play': 240, 'Indie': 233, 'Multiplay...  game   

                                                categories  \
10       [Multi-player, Valve Anti-Cheat enabled, Onlin...   
1000010  [Single-player, Partial Controller Support, St...   
1000030  [Multi-player, Single-player, Co-op, Steam Ach...   
1000080  [Multi-player, Single-player, Steam Achievemen...   
1000360  [Multi-player, Single-player, Shared/Split Scr...   
...                   

In [34]:
steam_games[steam_games.columns[9]].unique()

array(['10,000,000 .. 20,000,000', '200,000 .. 500,000',
       '100,000 .. 200,000', '500,000 .. 1,000,000',
       '2,000,000 .. 5,000,000', '5,000,000 .. 10,000,000',
       '1,000,000 .. 2,000,000', '20,000,000 .. 50,000,000',
       '50,000,000 .. 100,000,000', '200,000,000 .. 500,000,000'],
      dtype=object)

### > Done

## Убрать игры с нулевым онлайном

In [35]:
steam_games = steam_games[steam_games[steam_games.columns[15]] >= 100]
steam_games

appid                                  name  \
10            10                        Counter-Strike   
1003590  1003590              Tetris Effect: Connected   
1007040  1007040                 EARTH DEFENSE FORCE 5   
10090      10090            Call of Duty: World at War   
1009290  1009290  SWORD ART ONLINE Alicization Lycoris   
...          ...                                   ...   
996470    996470                   eFootball  PES 2020   
996580    996580               Spyro Reignited Trilogy   
997010    997010     Police Simulator: Patrol Officers   
997070    997070                     Marvel's Avengers   
99900      99900                        Spiral Knights   

                                         short_description  \
10       Play the world's number 1 online action game. ...   
1003590  Tetris® Effect: Connected adds an all-new robu...   
1007040  Stand and fight for humanity. This arcade shoo...   
10090    Call of Duty is back, redefining war like you'...   
1009290  Enjoy seamless, action-packed real-time battle...   
...                                                    ...   
996470   Experience unparalleled realism and authentici...   
996580   Same sick burns, same smoldering attitude, now...   
997010   Welcome to Brighton! Join the police force of ...   
997070   Assemble your team of Earth’s Mightiest Heroes...   
99900    Join the ranks of the Spiral Knights. Stranded...   

                                        developer  \
10                                          Valve   
1003590      Monstars Inc., Resonair, Stage Games   
1007040                                   SANDLOT   
10090                                    Treyarch   
1009290                          AQURIA Co., Ltd.   
...                                           ...   
996470               Konami Digital Entertainment   
996580                  Toys for Bob, Iron Galaxy   
997010                          Aesir Interactive   
997070   Crystal Dynamics, Eidos-Montréal, Nixxes   
99900                                 Grey Havens   

                            publisher  \
10                              Valve   
1003590                       Enhance   
1007040                   D3PUBLISHER   
10090                      Activision   
1009290    BANDAI NAMCO Entertainment   
...                               ...   
996470   Konami Digital Entertainment   
996580                     Activision   
997010         astragon Entertainment   
997070                    Square Enix   
99900                     Grey Havens   

                                                     genre  \
10                                                  Action   
1003590                                      Casual, Indie   
1007040                                             Action   
10090                                               Action   
1009290                             Action, Adventure, RPG   
...                                                    ...   
996470                                              Sports   
996580                                   Action, Adventure   
997010                            Simulation, Early Access   
997070                              Action, Adventure, RPG   
99900    Action, Adventure, Casual, Free to Play, Indie...   

                                                      tags  type  \
10       {'Action': 5426, 'FPS': 4831, 'Multiplayer': 3...  game   
1003590  {'Casual': 219, 'Puzzle': 214, 'Colorful': 186...  game   
1007040  {'Action': 169, 'Online Co-Op': 73, 'Mechs': 6...  game   
10090    {'Zombies': 1136, 'World War II': 848, 'FPS': ...  game   
1009290  {'RPG': 250, 'Action': 210, 'Adventure': 207, ...  game   
...                                                    ...   ...   
996470   {'Sports': 485, 'Soccer': 188, 'Multiplayer': ...  game   
996580   {'Adventure': 167, '3D Platformer': 161, 'Rema...  game   
997010   {'Simulation': 1045, 'Crime': 1040, 'Open Worl...  game   
997070   {'Act

### >Done

## > Следующим этапом будет: колонку тэги в датасете steam_games, которая записана в формате {'Action': 5426, 'FPS': 4831, ... } , привести к формату 'Action', 'FPS', ...

In [36]:
steam_games.columns

Index(['appid', 'name', 'short_description', 'developer', 'publisher', 'genre',
       'tags', 'type', 'categories', 'owners', 'positive', 'negative', 'price',
       'initialprice', 'discount', 'ccu', 'languages', 'platforms',
       'release_date', 'required_age', 'website', 'header_image'],
      dtype='object')

In [37]:
steam_games

appid                                  name  \
10            10                        Counter-Strike   
1003590  1003590              Tetris Effect: Connected   
1007040  1007040                 EARTH DEFENSE FORCE 5   
10090      10090            Call of Duty: World at War   
1009290  1009290  SWORD ART ONLINE Alicization Lycoris   
...          ...                                   ...   
996470    996470                   eFootball  PES 2020   
996580    996580               Spyro Reignited Trilogy   
997010    997010     Police Simulator: Patrol Officers   
997070    997070                     Marvel's Avengers   
99900      99900                        Spiral Knights   

                                         short_description  \
10       Play the world's number 1 online action game. ...   
1003590  Tetris® Effect: Connected adds an all-new robu...   
1007040  Stand and fight for humanity. This arcade shoo...   
10090    Call of Duty is back, redefining war like you'...   
1009290  Enjoy seamless, action-packed real-time battle...   
...                                                    ...   
996470   Experience unparalleled realism and authentici...   
996580   Same sick burns, same smoldering attitude, now...   
997010   Welcome to Brighton! Join the police force of ...   
997070   Assemble your team of Earth’s Mightiest Heroes...   
99900    Join the ranks of the Spiral Knights. Stranded...   

                                        developer  \
10                                          Valve   
1003590      Monstars Inc., Resonair, Stage Games   
1007040                                   SANDLOT   
10090                                    Treyarch   
1009290                          AQURIA Co., Ltd.   
...                                           ...   
996470               Konami Digital Entertainment   
996580                  Toys for Bob, Iron Galaxy   
997010                          Aesir Interactive   
997070   Crystal Dynamics, Eidos-Montréal, Nixxes   
99900                                 Grey Havens   

                            publisher  \
10                              Valve   
1003590                       Enhance   
1007040                   D3PUBLISHER   
10090                      Activision   
1009290    BANDAI NAMCO Entertainment   
...                               ...   
996470   Konami Digital Entertainment   
996580                     Activision   
997010         astragon Entertainment   
997070                    Square Enix   
99900                     Grey Havens   

                                                     genre  \
10                                                  Action   
1003590                                      Casual, Indie   
1007040                                             Action   
10090                                               Action   
1009290                             Action, Adventure, RPG   
...                                                    ...   
996470                                              Sports   
996580                                   Action, Adventure   
997010                            Simulation, Early Access   
997070                              Action, Adventure, RPG   
99900    Action, Adventure, Casual, Free to Play, Indie...   

                                                      tags  type  \
10       {'Action': 5426, 'FPS': 4831, 'Multiplayer': 3...  game   
1003590  {'Casual': 219, 'Puzzle': 214, 'Colorful': 186...  game   
1007040  {'Action': 169, 'Online Co-Op': 73, 'Mechs': 6...  game   
10090    {'Zombies': 1136, 'World War II': 848, 'FPS': ...  game   
1009290  {'RPG': 250, 'Action': 210, 'Adventure': 207, ...  game   
...                                                    ...   ...   
996470   {'Sports': 485, 'Soccer': 188, 'Multiplayer': ...  game   
996580   {'Adventure': 167, '3D Platformer': 161, 'Rema...  game   
997010   {'Simulation': 1045, 'Crime': 1040, 'Open Worl...  game   
997070   {'Act

In [38]:
for i in (steam_games[steam_games.columns[0]]):
    steam_games[steam_games.columns[6]][i] = ', '.join([str(key) for key in steam_games[steam_games.columns[6]][i].keys()][:4])


C:\Users\Dislis\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3437: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [39]:
steam_games

appid                                  name  \
10            10                        Counter-Strike   
1003590  1003590              Tetris Effect: Connected   
1007040  1007040                 EARTH DEFENSE FORCE 5   
10090      10090            Call of Duty: World at War   
1009290  1009290  SWORD ART ONLINE Alicization Lycoris   
...          ...                                   ...   
996470    996470                   eFootball  PES 2020   
996580    996580               Spyro Reignited Trilogy   
997010    997010     Police Simulator: Patrol Officers   
997070    997070                     Marvel's Avengers   
99900      99900                        Spiral Knights   

                                         short_description  \
10       Play the world's number 1 online action game. ...   
1003590  Tetris® Effect: Connected adds an all-new robu...   
1007040  Stand and fight for humanity. This arcade shoo...   
10090    Call of Duty is back, redefining war like you'...   
1009290  Enjoy seamless, action-packed real-time battle...   
...                                                    ...   
996470   Experience unparalleled realism and authentici...   
996580   Same sick burns, same smoldering attitude, now...   
997010   Welcome to Brighton! Join the police force of ...   
997070   Assemble your team of Earth’s Mightiest Heroes...   
99900    Join the ranks of the Spiral Knights. Stranded...   

                                        developer  \
10                                          Valve   
1003590      Monstars Inc., Resonair, Stage Games   
1007040                                   SANDLOT   
10090                                    Treyarch   
1009290                          AQURIA Co., Ltd.   
...                                           ...   
996470               Konami Digital Entertainment   
996580                  Toys for Bob, Iron Galaxy   
997010                          Aesir Interactive   
997070   Crystal Dynamics, Eidos-Montréal, Nixxes   
99900                                 Grey Havens   

                            publisher  \
10                              Valve   
1003590                       Enhance   
1007040                   D3PUBLISHER   
10090                      Activision   
1009290    BANDAI NAMCO Entertainment   
...                               ...   
996470   Konami Digital Entertainment   
996580                     Activision   
997010         astragon Entertainment   
997070                    Square Enix   
99900                     Grey Havens   

                                                     genre  \
10                                                  Action   
1003590                                      Casual, Indie   
1007040                                             Action   
10090                                               Action   
1009290                             Action, Adventure, RPG   
...                                                    ...   
996470                                              Sports   
996580                                   Action, Adventure   
997010                            Simulation, Early Access   
997070                              Action, Adventure, RPG   
99900    Action, Adventure, Casual, Free to Play, Indie...   

                                                      tags  type  \
10                       Action, FPS, Multiplayer, Shooter  game   
1003590                 Casual, Puzzle, Colorful, Abstract  game   
1007040  Action, Online Co-Op, Mechs, Third-Person Shooter  game   
10090              Zombies, World War II, FPS, Multiplayer  game   
1009290                      RPG, Action, Adventure, Anime  game   
...                                                    ...   ...   
996470           Sports, Soccer, Multiplayer, Free to Play  game   
996580           Adventure, 3D Platformer, Remake, Dragons  game   
997010          Simulation, Crime, Open World, Multiplayer  game   
997070        

### >Done

In [40]:
steam_games[['appid', 'tags', 'genre']].to_csv("steam_df.csv") #ВЫГРУЗКА ДАТАСЕТА

## Моделим

In [41]:
def NoGames(df):
    for i in df[df.columns[1]]:
        if list(steam_games[steam_games.columns[0]]).count(i) == 0:
            df.drop(df[df['appid'] == i].index, inplace=True)
    return df

In [42]:
our_user = [['YUZH1542', 381210, 0.8585413273568598], ['YUZH1542', 570, 0.846203955184673], ['YUZH1542', 1966720, 0.5502518907629605], ['YUZH1542', 359550, 0.21241669277825437], ['YUZH1542', 730, 0.17976154640026176], ['YUZH1542', 440, 0.13617110024733045], ['YUZH1542', 512900, 0.1322739297684637], ['YUZH1542', 304930, 0.10032414995869667], ['YUZH1542', 632360, 0.08891102030178387], ['YUZH1542', 312530, 0.08262687563499269], ['YUZH1542', 1172620, 0.07979501154360376], ['YUZH1542', 1621690, 0.07579544508048934], ['YUZH1542', 466240, 0.07093988675257049], ['YUZH1542', 438740, 0.06933252164027025], ['YUZH1542', 4000, 0.06834996102206033], ['YUZH1542', 945360, 0.06768699713777802], ['YUZH1542', 646570, 0.06419947719080635], ['YUZH1542', 544920, 0.06013752876966535], ['YUZH1542', 274190, 0.05815526314990437], ['YUZH1542', 739630, 0.05447175155226524]]
our_user = pd.DataFrame(our_user, columns=['user','appid', 'rating'])
NoGames(our_user)

user    appid    rating
0   YUZH1542   381210  0.858541
1   YUZH1542      570  0.846204
3   YUZH1542   359550  0.212417
4   YUZH1542      730  0.179762
5   YUZH1542      440  0.136171
6   YUZH1542   512900  0.132274
7   YUZH1542   304930  0.100324
8   YUZH1542   632360  0.088911
9   YUZH1542   312530  0.082627
10  YUZH1542  1172620  0.079795
11  YUZH1542  1621690  0.075795
12  YUZH1542   466240  0.070940
13  YUZH1542   438740  0.069333
14  YUZH1542     4000  0.068350
15  YUZH1542   945360  0.067687
16  YUZH1542   646570  0.064199
18  YUZH1542   274190  0.058155
19  YUZH1542   739630  0.054472

In [43]:
dset = pd.read_csv("steam_df.csv")
dset = dset.drop("Unnamed: 0", axis=1)
dset

appid                                               tags  \
0          10                  Action, FPS, Multiplayer, Shooter   
1     1003590                 Casual, Puzzle, Colorful, Abstract   
2     1007040  Action, Online Co-Op, Mechs, Third-Person Shooter   
3       10090            Zombies, World War II, FPS, Multiplayer   
4     1009290                      RPG, Action, Adventure, Anime   
...       ...                                                ...   
1640   996470          Sports, Soccer, Multiplayer, Free to Play   
1641   996580          Adventure, 3D Platformer, Remake, Dragons   
1642   997010         Simulation, Crime, Open World, Multiplayer   
1643   997070                  Action, Adventure, RPG, Superhero   
1644    99900  Free to Play, Massively Multiplayer, RPG, Adve...   

                                                  genre  
0                                                Action  
1                                         Casual, Indie  
2                                                Action  
3                                                Action  
4                                Action, Adventure, RPG  
...                                                 ...  
1640                                             Sports  
1641                                  Action, Adventure  
1642                           Simulation, Early Access  
1643                             Action, Adventure, RPG  
1644  Action, Adventure, Casual, Free to Play, Indie...  

[1645 rows x 3 columns]

In [44]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel

tfidf = TfidfVectorizer(stop_words='english')
data_matrix = tfidf.fit_transform(dset['tags'])

# Построение матрицы схожести косинусного расстояния
cosine_similarities = linear_kernel(data_matrix, data_matrix)

# Функция для получения рекомендаций
def get_recommendations(user_games):
    # Индексы игр, которые пользователь уже играл
    indices = [dset[dset['appid'] == game_id].index[0] for game_id in user_games]

    # Суммируем схожести всех игр, в которые играл пользователь
    similarity_scores = cosine_similarities[indices].sum(axis=0)

    # Сортировка схожих игр по убыванию
    top_indices = similarity_scores.argsort()[::-1]

    # Исключаем игры, которые пользователь уже играл
    top_indices = [index for index in top_indices if index not in indices]

    # Получаем список рекомендованных игр
    recommendations = dset.iloc[top_indices[:10]]['appid']

    return recommendations

# Пример использования
user_games = list(our_user[our_user.columns[1]]) # Замени это на список реальных игр, в которые играл пользователь
recommended_games = list(get_recommendations(user_games))


In [45]:
print(recommended_games)

[223710, 744900, 1274570, 1361000, 967050, 509980, 1056600, 1549180, 915320, 700330]


## Датасет 2

In [208]:
users_df = pd.read_csv("steam.csv")
users_df

user   appid    rating
0      smaile2003     570  0.940148
1      smaile2003  431960  0.572062
2      smaile2003     730  0.170746
3      smaile2003     440  0.143814
4      smaile2003    4000  0.120521
...           ...     ...       ...
10006   yunggface  611790  0.086381
10007   yunggface  960090  0.083394
10008   yunggface   15100  0.080296
10009   yunggface  495140  0.070186
10010   yunggface  292030  0.069146

[10011 rows x 3 columns]

In [209]:
NoGames(users_df)

user   appid    rating
0      smaile2003     570  0.940148
1      smaile2003  431960  0.572062
2      smaile2003     730  0.170746
3      smaile2003     440  0.143814
4      smaile2003    4000  0.120521
...           ...     ...       ...
10005   yunggface  582500  0.087215
10006   yunggface  611790  0.086381
10007   yunggface  960090  0.083394
10008   yunggface   15100  0.080296
10010   yunggface  292030  0.069146

[7906 rows x 3 columns]

### Отсутствующие игры убраны

## Моделим 2

### merge users_df 
### --в сет сунуть пользователя
### убрать повторения
### сделать играм avg rating


In [210]:
#dataset только из оцененных игр

In [213]:
ratings_count = pd.DataFrame(columns=["user", "rating_counter"])
ratings_count = users_df.groupby('appid')['user'].count() > 10
ratings_count = ratings_count.loc[ratings_count]
rating_viable = ratings_count.index.tolist()
len(rating_viable)

161

In [214]:
users_df = users_df[users_df['appid'].isin(rating_viable)]
users_df

user    appid    rating
0      smaile2003      570  0.940148
1      smaile2003   431960  0.572062
2      smaile2003      730  0.170746
3      smaile2003      440  0.143814
4      smaile2003     4000  0.120521
...           ...      ...       ...
10001   yunggface   286160  0.097417
10003   yunggface  1063730  0.096295
10004   yunggface   550650  0.094011
10007   yunggface   960090  0.083394
10010   yunggface   292030  0.069146

[5643 rows x 3 columns]

In [215]:
users_df = pd.concat([users_df, our_user], ignore_index=True)
users_df = users_df.drop_duplicates()
users_df

user   appid    rating
0     smaile2003     570  0.940148
1     smaile2003  431960  0.572062
2     smaile2003     730  0.170746
3     smaile2003     440  0.143814
4     smaile2003    4000  0.120521
...          ...     ...       ...
5656    YUZH1542    4000  0.068350
5657    YUZH1542  945360  0.067687
5658    YUZH1542  646570  0.064199
5659    YUZH1542  274190  0.058155
5660    YUZH1542  739630  0.054472

[5496 rows x 3 columns]

In [216]:
final_df = users_df.merge(steam_games, on='appid')
final_df

user    appid    rating              name  \
0     smaile2003      570  0.940148            Dota 2   
1       etoneol9      570  0.964711            Dota 2   
2     nikitalyan      570  0.817905            Dota 2   
3        SonDVan      570  0.598754            Dota 2   
4          2rage      570  0.500000            Dota 2   
...          ...      ...       ...               ...   
5491    YUZH1542   512900  0.132274  Streets of Rogue   
5492    YUZH1542   312530  0.082627         Duck Game   
5493    YUZH1542  1621690  0.075795       Core Keeper   
5494    YUZH1542   646570  0.064199    Slay the Spire   
5495    YUZH1542   274190  0.058155          Broforce   

                                      short_description          developer  \
0     Every day, millions of players worldwide enter...              Valve   
1     Every day, millions of players worldwide enter...              Valve   
2     Every day, millions of players worldwide enter...              Valve   
3     Every day, millions of players worldwide enter...              Valve   
4     Every day, millions of players worldwide enter...              Valve   
...                                                 ...                ...   
5491  Fight, sneak, and hack your way through random...     Matt Dabrowski   
5492  Enter the futuristic year of 1984, an age wher...  Landon Podbielski   
5493  Explore an endless cavern of creatures, relics...           Pugstorm   
5494  We fused card games and roguelikes together to...    Mega Crit Games   
5495  When evil threatens the world, the world calls...         Free Lives   

             publisher                                        genre  \
0                Valve               Action, Free to Play, Strategy   
1                Valve               Action, Free to Play, Strategy   
2                Valve               Action, Free to Play, Strategy   
3                Valve               Action, Free to Play, Strategy   
4                Valve               Action, Free to Play, Strategy   
...                ...                                          ...   
5491         tinyBuild                Action, Adventure, Indie, RPG   
5492  Adult Swim Games                                Action, Indie   
5493   Fireshine Games  Adventure, Casual, Indie, RPG, Early Access   
5494   Mega Crit Games                              Indie, Strategy   
5495  Devolver Digital             Action, Adventure, Casual, Indie   

                                                   tags  type  ... price  \
0             Free to Play, MOBA, Multiplayer, Strategy  game  ...     0   
1             Free to Play, MOBA, Multiplayer, Strategy  game  ...     0   
2             Free to Play, MOBA, Multiplayer, Strategy  game  ...     0   
3             Free to Play, MOBA, Multiplayer, Strategy  game  ...     0   
4             Free to Play, MOBA, Multiplayer, Strategy  game  ...     0   
...                                                 ...   ...  ...   ...   
5491  Rogue-lite, RPG, Action Roguelike, Pixel Graphics  game  ...  1999   
5492  Multiplayer, Funny, Pixel Graphics, Local Mult...  game  ...  1299   
5493  Early Access, Sandbox, Pixel Graphics, Open Wo...  game  ...  1299   
5494  Roguelike Deckbuilder, Card Game, Card Battler...  game  ...   849   
5495             America, Action, Pixel Graphics, Co-op  game  ...  1499   

     initialprice discount     ccu  \
0               0        0  852995   
1               0        0  852995   
2               0        0  852995   
3               0        0  852995   
4               0        0  852995   
...           ...      ...     ...   
5491         1999        0     454   
5492         1299        0     257   
5493         1299        0    2413   
5494         2499       66    8371   
5495         1499        0     248   

                                              languages  \
0     English, Bulgarian, Czech, Danish, Dutch, Finn...   
1     English, Bulgarian, Czech, Danish, Dutch, F

In [217]:
print(len(list(final_df[final_df.columns[3]].unique())))
print(len(list(final_df[final_df.columns[0]].unique())))

166
490


In [218]:
df_grouped = final_df.groupby('appid')['rating'].mean()
df_grouped = steam_games.merge(df_grouped, on='appid')
df_grouped

appid                               name  \
0         10                     Counter-Strike   
1    1030840          Mafia: Definitive Edition   
2    1046930                    Dota Underlords   
3     105600                           Terraria   
4    1063730                          New World   
..       ...                                ...   
161     8930         Sid Meier's Civilization V   
162   945360                           Among Us   
163   960090                        Bloons TD 6   
164   976310                   Mortal Kombat 11   
165   976730  Halo: The Master Chief Collection   

                                     short_description  \
0    Play the world's number 1 online action game. ...   
1    An inadvertent brush with the mob thrusts cabd...   
2    Hire a crew and destroy your rivals in this ne...   
3    Dig, fight, explore, build! Nothing is impossi...   
4    Explore a thrilling, open-world MMO filled wit...   
..                                                 ...   
161  Create, discover, and download new player-crea...   
162  An online and local party game of teamwork and...   
163  The Bloons are back and better than ever! Get ...   
164  Mortal Kombat is back and better than ever in ...   
165  The Master Chief’s iconic journey includes six...   

                                             developer  \
0                                                Valve   
1                                            Hangar 13   
2                                                Valve   
3                                             Re-Logic   
4                                         Amazon Games   
..                                                 ...   
161          Firaxis Games, Aspyr (Mac), Aspyr (Linux)   
162                                         Innersloth   
163                                         Ninja Kiwi   
164                  NetherRealm Studios, QLOC, Shiver   
165  343 Industries, Splash Damage, Ruffian Games, ...   

                                             publisher  \
0                                                Valve   
1                                                   2K   
2                                                Valve   
3                                             Re-Logic   
4                                         Amazon Games   
..                                                 ...   
161                     2K, Aspyr (Mac), Aspyr (Linux)   
162                                         Innersloth   
163                                         Ninja Kiwi   
164  Warner Bros. Games, Warner Bros. Interactive E...   
165                                  Xbox Game Studios   

                                             genre  \
0                                           Action   
1                                Action, Adventure   
2                   Casual, Free to Play, Strategy   
3                    Action, Adventure, Indie, RPG   
4    Action, Adventure, Massively Multiplayer, RPG   
..                                             ...   
161                                       Strategy   
162                                         Casual   
163                                       Strategy   
164                                         Action   
165                                         Action   

                                                  tags  type  \
0                    Action, FPS, Multiplayer, Shooter  game   
1                Action, Open World, Story Rich, Crime  game   
2         Free to Play, Strategy, Casual, Auto Battler  game   
3     Open World Survival Craft, Sandbox, Survival, 2D  game   
4       Massively Multiplayer, Open World, MMORPG, RPG  game   
..                                                 ...   ...   
161  Turn-Based Strategy, Strategy, Turn-Based, Mul...  game   
162  Multiplayer, Online Co-Op, Social Deduction, S...  game   
163  Tower Defense, Strategy, Multiplayer, Singlepl...  game   
164           

In [206]:
df_temp = df_grouped.drop("rating", axis=1)
users = users_df['user'].unique()
all_combinations = pd.DataFrame({'user': users})
all_combinations['key'] = 0
df_temp['key'] = 0
result = pd.merge(all_combinations, df_temp, on='key', how='outer').drop('key', axis=1)

# Добавление оценок пользователей к созданному датасету
result = pd.merge(result, users_df, on=['user', 'appid'], how='left')
result['rating'] = result['rating'].fillna(0)

result

user    appid                               name  \
0      smaile2003       10                     Counter-Strike   
1      smaile2003   105600                           Terraria   
2      smaile2003  1063730                          New World   
3      smaile2003   107410                             Arma 3   
4      smaile2003  1085660                          Destiny 2   
...           ...      ...                                ...   
51841    YUZH1542   892970                            Valheim   
51842    YUZH1542     8930         Sid Meier's Civilization V   
51843    YUZH1542   945360                           Among Us   
51844    YUZH1542   960090                        Bloons TD 6   
51845    YUZH1542   976730  Halo: The Master Chief Collection   

                                       short_description  \
0      Play the world's number 1 online action game. ...   
1      Dig, fight, explore, build! Nothing is impossi...   
2      Explore a thrilling, open-world MMO filled wit...   
3      Experience true combat gameplay in a massive m...   
4      Destiny 2 is an action MMO with a single evolv...   
...                                                  ...   
51841  A brutal exploration and survival game for 1-1...   
51842  Create, discover, and download new player-crea...   
51843  An online and local party game of teamwork and...   
51844  The Bloons are back and better than ever! Get ...   
51845  The Master Chief’s iconic journey includes six...   

                                               developer  \
0                                                  Valve   
1                                               Re-Logic   
2                                           Amazon Games   
3                                    Bohemia Interactive   
4                                                 Bungie   
...                                                  ...   
51841                                       Iron Gate AB   
51842          Firaxis Games, Aspyr (Mac), Aspyr (Linux)   
51843                                         Innersloth   
51844                                         Ninja Kiwi   
51845  343 Industries, Splash Damage, Ruffian Games, ...   

                            publisher  \
0                               Valve   
1                            Re-Logic   
2                        Amazon Games   
3                 Bohemia Interactive   
4                              Bungie   
...                               ...   
51841         Coffee Stain Publishing   
51842  2K, Aspyr (Mac), Aspyr (Linux)   
51843                      Innersloth   
51844                      Ninja Kiwi   
51845               Xbox Game Studios   

                                               genre  \
0                                             Action   
1                      Action, Adventure, Indie, RPG   
2      Action, Adventure, Massively Multiplayer, RPG   
3                       Action, Simulation, Strategy   
4                    Action, Adventure, Free to Play   
...                                              ...   
51841    Action, Adventure, Indie, RPG, Early Access   
51842                                       Strategy   
51843                                         Casual   
51844                                       Strategy   
51845                                         Action   

                                                    tags  type  \
0                      Action, FPS, Multiplayer, Shooter  game   
1       Open World Survival Craft, Sandbox, Survival, 2D  game   
2         Massively Multiplayer, Open World, MMORPG, RPG  game   
3              Action, Military, Multiplayer, Simulation  game   
4          Free to Play, Open World, FPS, Looter Shooter  game   
...                                                  ...   ...   
51841  Open World Survival Craft, Survival, Online Co...  game   
51842  Turn-Based Strategy, Strategy, Turn-Based, Mul...  game   
51843  Multiplayer, Online Co-

In [246]:
#from surprise import Reader, Dataset, SVD
import surprise

ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject

In [224]:
reader = Reader(rating_scale=(1, 10))
data   = Dataset.load_from_df(result[['userID','bookTitle','bookRating']], reader)
trainset = data.build_full_trainset()
model = SVD(n_factors=50, n_epochs=10, lr_all=0.005, reg_all= 0.2)
model.fit(trainset)
testset = trainset.build_anti_testset()
predictions = model.test(testset)
predictions_df = pd.DataFrame(predictions)

NameError: name 'Reader' is not defined

In [ ]:
def generate_recommendationsSVD(user='smaile2003', get_recommend =10):
    
    # get the top get_recommend predictions for userID
    
    predictions_userID = predictions_df[predictions_df['user'] == user].\
                         sort_values(by="est", ascending = False).head(get_recommend)
    recommendations = []
    recommendations.append(list(predictions_userID['user']))
    recommendations = recommendations[0]
    rec_games = df_grouped[df_grouped.appid.isin(recommendations)]
    return rec_games


In [ ]:
recommendationsSVD = generate_recommendationsSVD(user='YUZH1542', get_recommend =10)